src/Miunie.Discord/Impersonation.cs

```diff
@@ -31,41 +31,48 @@ public IEnumerable<GuildView> GetAvailableGuilds()
                 Name = g.Name
             });
 
-        public async Task<IEnumerable<TextChannelView>> GetAvailableTextChannelsAsync(ulong guildId)
+        public IEnumerable<TextChannelView> GetAvailableTextChannelsAsync(ulong guildId)
         {
             if(guildId == 0) { return new TextChannelView[0]; }
 
             var guild = _discord.Client.GetGuild(guildId);
-            var textChannels = guild.Channels.Where(c => c is SocketTextChannel).Cast<SocketTextChannel>();
-            var result = new List<TextChannelView>();
-            foreach (var channel in textChannels)
-            {
-                try
-                {
-                    result.Add(new TextChannelView
-                    {
-                        Id = channel.Id,
-                        Name = $"# {channel.Name}",
-                        Messages = await GetMessagesFrom(channel)
-                    });
-                }
-                catch (Exception)
+            var textChannels = guild.Channels
+                .Where(c => c is SocketTextChannel && c.GetUser(_discord.Client.CurrentUser.Id).GetPermissions(c).ViewChannel)
+                .Cast<SocketTextChannel>()
+                .Select(channel => new TextChannelView
                 {
-                    _logger.Log($"Miunie cannot read from the '{channel.Name}' channel.");
-                }
-            }
+                    Id = channel.Id,
+                    Name = $"# {channel.Name}",
+                    Messages = new MessageView[0]
+                });
```

Comment:
So if we're doing this, I'd suggest extracting the content of the `Where` clause into a method as well as the contents of this `Select` clause.

Something like
`.Where(IsViewableTextChannel)`
and
`Select(ToTextChannelView)`

That would read really well! 😊